main.tf

```diff
@@ -2,92 +2,163 @@ resource "aws_cloudwatch_metric_alarm" "httpcode_target_5xx_count" {
-    "TargetGroup"  = var.target_group_id
-    "LoadBalancer" = var.load_balancer_id
+  metric_query {
+    id          = "e1"
+    expression  = length(var.metric_expression) > 0 ? var.metric_expression : "FILL(m1, LINEAR)"
+    return_data = (length(var.metric_expression) > 0 ? "true" : "false")
+  }
+
+  metric_query {
+    id = "m1"
+    return_data = (length(var.metric_expression) > 0 ? "false" : "true")
+
+    metric {
+      metric_name         = "HTTPCode_Target_5XX_Count"
+      namespace           = "AWS/ApplicationELB"
+      period              = var.statistic_period
+      stat                = "Sum"
+
+      dimensions = {
+        "TargetGroup"  = var.target_group_id
+        "LoadBalancer" = var.load_balancer_id
+      }
+    }
   }
 }
 
 resource "aws_cloudwatch_metric_alarm" "httpcode_lb_5xx_count" {
   alarm_name          = "${var.prefix}alb-${var.load_balancer_id}-high5XXCount"
   comparison_operator = "GreaterThanThreshold"
   evaluation_periods  = var.evaluation_period
-  metric_name         = "HTTPCode_ELB_5XX_Count"
-  namespace           = "AWS/ApplicationELB"
-  period              = var.statistic_period
-  statistic           = "Sum"
   threshold           = "0"
   alarm_description   = "Average API 5XX load balancer error code count is too high"
   alarm_actions       = var.actions_alarm
   ok_actions          = var.actions_ok
 
-  dimensions = {
-    "LoadBalancer" = var.load_balancer_id
+  metric_query {
+    id          = "e1"
+    expression  = length(var.metric_expression) > 0 ? var.metric_expression : "FILL(m1, LINEAR)"
+    return_data = (length(var.metric_expression) > 0 ? "true" : "false")
+  }
+
+  metric_query {
+    id = "m1"
+    return_data = (length(var.metric_expression) > 0 ? "false" : "true")
+
+    metric {
+      metric_name         = "HTTPCode_ELB_5XX_Count"
+      namespace           = "AWS/ApplicationELB"
+      period              = var.statistic_period
+      stat                = "Sum"
+
+      dimensions = {
+        "LoadBalancer" = var.load_balancer_id
+      }
+    }
   }
 }
 
 resource "aws_cloudwatch_metric_alarm" "target_response_time_average" {
   alarm_name          = "${var.prefix}alb-tg-${var.target_group_id}-highResponseTime"
   comparison_operator = "GreaterThanThreshold"
   evaluation_periods  = var.evaluation_period
-  metric_name         = "TargetResponseTime"
-  namespace           = "AWS/ApplicationELB"
-  period              = var.statistic_period
-  statistic           = "Average"
   threshold           = var.response_time_threshold
   alarm_description   = format("Average API response time is greater than %s", var.response_time_threshold)
   alarm_actions       = var.actions_alarm
   ok_actions          = var.actions_ok
 
-  dimensions = {
-    "TargetGroup"  = var.target_group_id
-    "LoadBalancer" = var.load_balancer_id
+  metric_query {
+    id          = "e1"
+    expression  = length(var.metric_expression) > 0 ? var.metric_expression : "FILL(m1, LINEAR)"
```

Comment:
so effectively `FILL(m1, LINEAR)` is now the default? This is not what I meant. 

if var.metric_expression is empty, then the alarm should not just use the raw metric value

Reply:
yes, please, lets first apply without the expression and apply with one.

Reply:
Using `FILL(m1, REPEAT)` for `metric_expression` and setting `treat_missing_data` to `ignore` appears to fix the flapping problem.

The limitations with this approach is that we won't receive notifications if a target group completely goes away and therefore has missing data.